jupyter-extension/src/plugins/mount/components/Datum/hooks/useDatum.tsx

```diff
@@ -58,37 +68,92 @@ export const useDatum = (
         idx: currentDatumInfo.curr_idx,
         num_datums: currentDatumInfo.num_datums,
       });
-      setInputSpec(YAML.stringify(currentDatumInfo.input, null, 2));
+      setInputSpec(inputSpecObjToText(currentDatumInfo.input));
     }
   }, [showDatum]);
 
+  useEffect(() => {
+    if (typeof datumViewInputSpec === 'string') {
+      setInputSpec(datumViewInputSpec);
+    } else {
+      let specToShow = {};
+      if (Object.keys(datumViewInputSpec).length === 0) {
+        specToShow = repoViewInputSpec;
+      } else {
+        specToShow = datumViewInputSpec;
+      }
+      setInputSpec(inputSpecObjToText(specToShow));
+      setInitialInputSpec(specToShow);
+    }
+  }, [showDatum, repoViewInputSpec]);
+
+  const saveInputSpec = (): void => {
+    try {
+      const inputSpecObj = inputSpecTextToObj();
+      if (isEqual(repoViewInputSpec, inputSpecObj)) {
+        setDatumViewInputSpec({});
+      } else {
+        setDatumViewInputSpec(inputSpecObj ? inputSpecObj : {});
+      }
+    } catch (e) {
+      if (e instanceof YAML.YAMLParseError) {
+        setDatumViewInputSpec(inputSpec);
+      } else {
+        throw e;
+      }
+    }
+  };
+
+  let useYaml = isYaml;
+  const inputSpecTextToObj = (): JSONObject => {
+    let spec = {};
+    try {
+      spec = JSON.parse(inputSpec);
+      useYaml = false;
+      setIsYaml(false);
+    } catch (e) {
+      if (e instanceof SyntaxError) {
+        spec = YAML.parse(inputSpec);
+        useYaml = true;
+        setIsYaml(true);
+      } else {
+        throw e;
+      }
+    }
+    return spec;
+  };
+
+  useEffect(() => {
```

Comment:
@pleeko Is there a better approach I can take here? 
The issue is the useState function `setIsYaml` is async and I'm unable to `await` it, so I set `isYaml` on lines 112 and 117 to instantly reflect whether the input spec we're processing is yaml or not. I needed this info because of the logic in the `inputSpecObjToText()` function on line 130. And then I keep this in a state variable to be used later when the user switches back into datum mode.